File: src/main/java/idlezoo/game/domain/Zoo.java
package idlezoo.game.domain;

import com.fasterxml.jackson.annotation.JsonPropertyOrder;
import idlezoo.game.domain.Perks.Perk;
import org.immutables.value.Value;

import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Value.Immutable
@JsonPropertyOrder(alphabetic = true)
public abstract class Zoo {

    private Map<String, ZooBuildings> buildingsMap;

    public double getMoneyIncome() {
        return getBaseIncome() + getPerkIncome();
    }

    public ZooBuildings animal(String animal) {
        if (buildingsMap == null) {
            buildingsMap = buildingsMap(getBuildings());
        }
        return buildingsMap.get(animal);
    }

    public abstract List<ZooBuildings> getBuildings();

    public abstract String getName();

    public abstract double getMoney();

    public abstract double getPerkIncome();

    public abstract List<Perk> getPerks();

    public abstract List<Perk> getAvailablePerks();

    public abstract int getFightWins();

    public abstract int getFightLosses();

    public abstract boolean isWaitingForFight();

    public abstract long getChampionTime();

    public abstract double getBaseIncome();

    private static Map<String, ZooBuildings> buildingsMap(List<ZooBuildings> buildings) {
        Map<String, ZooBuildings> result = new HashMap<>();
        for (ZooBuildings building : buildings) {
            result.put(building.getName(), building);
        }
        return Collections.unmodifiableMap(result);
    }
}

File: src/main/kotlin/idlezoo/game/domain/Zoo.kt
package idlezoo.game.domain;

import com.fasterxml.jackson.annotation.JsonPropertyOrder;
import idlezoo.game.domain.Perks.Perk;

@JsonPropertyOrder(alphabetic = true)
data class Zoo(
        val buildings: List<ZooBuildings>,
        val name: String,
        val money: Double,
        val perkIncome: Double,
        val perks: List<Perk>,
        val availablePerks: List<Perk>,
        val fightWins: Int,
        val fightLosses: Int,
        val waitingForFight: Boolean,
        val championTime: Long,
        val baseIncome: Double
) {

    private val buildingsMap: Map<String, ZooBuildings>

    init {
        buildingsMap = mutableMapOf()
        buildings.forEach { buildingsMap.put(it.getName(), it) }
    }

    fun getMoneyIncome(): Double = baseIncome + perkIncome
    fun animal(animal: String): ZooBuildings? = buildingsMap[animal]

}